File: adapters/django/src/fieldseal_django/context.py
"""Tenant binding and per-column `FieldContext` assembly (docs/12 §4).

Django field types cannot see the record they belong to. `get_db_prep_value`
receives a value, a connection and a `prepared` flag -- not the instance, not
the row, not the request. Spec §10's L3 row for Django therefore reads
"documented contextvar side channel", and this module is that side channel,
written to fail closed.

**The failure mode this is built around.** A tenant-bound column whose tenant
is unset must not quietly encrypt under a tenantless context: the write would
succeed, the row would be unreadable by every correctly-configured reader,
and nothing would say so until someone tried to read it. `require_tenant()`
raises instead. That is spec §10.2's rule applied to context rather than to
queries.
"""

from __future__ import annotations

from collections.abc import Iterator
from contextlib import contextmanager
from contextvars import ContextVar, Token

from fieldseal import FieldContext

from .errors import FieldsealConfigurationError

#: Set per request (middleware), per task (Celery), or per command. A
#: ContextVar rather than thread-local so that async views and thread pools
#: inherit it the way Django's own async plumbing expects.
_tenant: ContextVar[bytes | None] = ContextVar(
    "fieldseal_django_tenant", default=None)


def set_tenant(tenant_id: bytes | str | None) -> Token[bytes | None]:
    """Bind the tenant for the current context.

    Returns the `Token` so a caller can restore the previous value; prefer
    `tenant_scope()`, which does it for you.
    """
    if tenant_id is None:
        return _tenant.set(None)
    if isinstance(tenant_id, str):
        tenant_id = tenant_id.encode("utf-8")
    if not isinstance(tenant_id, bytes):
        raise FieldsealConfigurationError(
            f"tenant_id must be bytes or str, got {type(tenant_id).__name__}"
        )
    return _tenant.set(tenant_id)


def get_tenant() -> bytes | None:
    return _tenant.get()


@contextmanager
def tenant_scope(tenant_id: bytes | str | None) -> Iterator[None]:
    """Bind a tenant for the duration of a block, restoring the previous one.

    The shape management commands and Celery tasks should use, because the
    middleware that would otherwise set it never runs there.
    """
    token = set_tenant(tenant_id)
    try:
        yield
    finally:
        _tenant.reset(token)


def require_tenant(model: str, field: str) -> bytes:
    """The tenant for a column declared `tenant_bound`, or a refusal.

    Failing closed here is the whole point. The alternative -- treating an
    unset tenant as "no tenant" -- writes a row under a context no configured
    reader will reconstruct, and the error surfaces later as
    `COMMITMENT_INVALID` on a read, which names neither the column nor the
    missing binding.
    """
    tenant = _tenant.get()
    if tenant is None:
        raise FieldsealConfigurationError(
            f"{model}.{field} is declared tenant_bound and no tenant is set "
            "in this context. Set one with "
            "`fieldseal_django.context.set_tenant(...)` or the "
            "`tenant_scope(...)` context manager -- the shipped middleware "
            "covers requests, but management commands, Celery tasks and "
            "shell sessions run outside it and must set it themselves. "
            "Encrypting without it would store a row that no correctly "
            "configured reader can decrypt."
        )
    return tenant


def build_context(
    *,
    table_uuid: bytes,
    column_uuid: bytes,
    tenant_bound: bool,
    model: str,
    field: str,
    purpose: str = "encrypt",
) -> FieldContext:
    """Assemble the per-operation `FieldContext` for a column.

    `row_id` is always `None` in v0 (`docs/12` §4, §9): Django cannot see the
    primary key at INSERT time for identity keys, so L3-row is deferred
    rather than half-implemented. The coverage matrix records it as ❌ with
    that reason rather than leaving it unmentioned.
    """
    return FieldContext(
        table_uuid=table_uuid,
        column_uuid=column_uuid,
        purpose=purpose,
        tenant_id=require_tenant(model, field) if tenant_bound else None,
        row_id=None,
    )

File: adapters/prisma/src/context.ts
/**
 * Tenant binding and context assembly (spec §10, L3).
 *
 * The Prisma extension runs before the query engine acquires a connection and
 * sees only `(model, operation, args)` -- never the record. So the tenant
 * arrives out of band, exactly as it does in Django, and spec §10.1 marks
 * Prisma's L3 "⚠️ partial" for this reason. Two documented side channels:
 *
 *  - `tenantScope(id, fn)` -- an `AsyncLocalStorage`, set by request
 *    middleware. Survives `await` across the whole request, which a
 *    thread-local could not.
 *  - the `tenant` option -- a callback given `(args, model, operation)`, for
 *    deployments that carry the tenant in the arguments themselves.
 *
 * **Fail-closed.** A tenant-bound column with no resolvable tenant refuses the
 * write. Falling back to a tenantless context would store a row that no
 * correctly configured reader can decrypt -- and because spec §6.3 binds the
 * context into the derived key, the reader's error would be
 * `COMMITMENT_INVALID`: a decrypt-side failure reported for a write-side
 * configuration mistake, arbitrarily far from the cause.
 *
 * `rowId` is always `null`: L3-row is not in v0 (`docs/13` §8). The extension
 * runs before the query, so a database-generated id does not exist yet.
 */

import { AsyncLocalStorage } from "node:async_hooks";

import type { FieldContext } from "@fieldseal/core";

import { FieldsealConfigurationError } from "./errors.ts";
import type { EncryptedFieldDecl, ResolvedModel } from "./fieldmap.ts";

const storage = new AsyncLocalStorage<Uint8Array>();

/** Run `fn` with `tenant` bound for every fieldseal operation inside it. */
export function tenantScope<T>(tenant: Uint8Array | string, fn: () => T): T {
  return storage.run(toBytes(tenant), fn);
}

/** The tenant bound in this async context, or `null`. */
export function getTenant(): Uint8Array | null {
  return storage.getStore() ?? null;
}

function toBytes(v: Uint8Array | string): Uint8Array {
  return typeof v === "string" ? Buffer.from(v, "utf8") : v;
}

/** Resolves the tenant for one operation. */
export type TenantResolver = (
  args: unknown,
  model: string,
  operation: string,
) => Uint8Array | string | null | undefined;

export interface ContextOptions {
  readonly resolver?: TenantResolver | undefined;
  /**
   * Told about every `FieldContext` this module builds (L4, `warm.ts`).
   *
   * This module is the single place a context is constructed -- `buildContext`
   * makes the value context and `indexContext` derives the index one from it --
   * so recording here records all of them. That is what makes `warm()`-on-miss
   * able to name the keys an operation needs without a second walk of the
   * arguments: the passes themselves report what they asked for.
   */
  readonly record?: ((ctx: FieldContext) => void) | undefined;
}

/**
 * Assemble the `FieldContext` for one column.
 *
 * The purpose is `"encrypt"`. `FieldContext` is a plain interface in this core,
 * not a class, so there is no `forIndex` method to call: `indexContext()` below
 * derives the index context from this one by replacing the purpose, which keeps
 * the two from ever disagreeing about table, column or tenant.
 */
export function buildContext(
  model: ResolvedModel,
  field: EncryptedFieldDecl,
  args: unknown,
  operation: string,
  opts: ContextOptions,
): FieldContext {
  // Only models with declarations reach here, and the generator refuses a
  // declared model without a table_uuid; a null here is an edited or stale map.
  if (model.tableUuid === null) {
    throw new FieldsealConfigurationError(
      `fieldseal: ${model.model}.${field.field} is declared encrypted but the ` +
        `field map carries no table_uuid for ${model.model}. The generator ` +
        `emits both together, so the map is stale or was edited -- re-run ` +
        `\`prisma generate\`.`,
    );
  }
  const tenant = field.tenantBound
    ? requireTenant(model, field, args, operation, opts)
    : null;
  const ctx: FieldContext = {
    tableUuid: uuidBytes(model.tableUuid),
    columnUuid: uuidBytes(field.columnUuid),
    tenantId: tenant,
    rowId: null,
    purpose: "encrypt",
  };
  opts.record?.(ctx);
  return ctx;
}

function requireTenant(
  model: ResolvedModel,
  field: EncryptedFieldDecl,
  args: unknown,
  operation: string,
  opts: ContextOptions,
): Uint8Array {
  const fromCallback = opts.resolver?.(args, model.model, operation);
  const tenant =
    fromCallback === null || fromCallback === undefined ? getTenant() : toBytes(fromCallback);
  if (tenant === null) {
    throw new FieldsealConfigurationError(
      `${model.model}.${field.field} is declared tenant_bound and no tenant is ` +
        `resolvable for this ${operation}. Set one with \`tenantScope(id, fn)\` ` +
        `-- request middleware covers requests, but scripts, queue workers and ` +
        `REPL sessions run outside it and must set it themselves -- or supply a ` +
        `\`tenant\` callback to fieldsealExtension(). Encrypting without it ` +
        `would store a row that no correctly configured reader can decrypt, and ` +
        `spec §6.3 binds the context into the derived key, so the reader would ` +
        `see COMMITMENT_INVALID rather than anything naming this cause.`,
    );
  }
  return tenant;
}

/**
 * The index-key context for a column, derived from its value context.
 *
 * Spec §5.2 makes the index key a *sibling* of the tenant DEK rather than a
 * child of it, and spec §6.1's purpose grammar is what selects between them.
 * Deriving this from the value context rather than rebuilding it is deliberate:
 * a table, column or tenant that differed between the two would derive an
 * index value under one identity and a ciphertext under another, and nothing
 * would raise -- the lookup would simply stop finding the row.
 */
export function indexContext(
  ctx: FieldContext,
  indexId: string,
  opts: ContextOptions = {},
): FieldContext {
  const derived: FieldContext = { ...ctx, purpose: `index:${indexId}` };
  // Recorded separately from the value context it came from: spec §5.2 makes
  // the index key a *sibling* of the tenant DEK rather than a child, so a
  // provider that warmed only the value context would leave every indexed
  // lookup stalled -- the failure Django's warm tests name as "a slow query
  // rather than a cold cache".
  opts.record?.(derived);
  return derived;
}

/** 8-4-4-4-12 hex -> the 16 bytes spec §6.1 derives from. */
export function uuidBytes(uuid: string): Uint8Array {
  return Buffer.from(uuid.replace(/-/g, ""), "hex");
}
